c10/core/DispatchKey.h

```diff
@@ -19,47 +27,85 @@ enum class DispatchKey : uint8_t {
+  XLATensorId,    // lives out of tree at https://github.com/pytorch/xla
+
+  // These are Caffe2 device types which we grandfathered into
+  // DispatchKey.
+  // TODO: Caffe2-only DispatchKeys actually should be removed from this enum
+  // and just simply be undispatchable.
+  MKLDNNTensorId, // (MKLDNN is treated as another "device" in Caffe2)
+  OpenGLTensorId,
+  OpenCLTensorId,
+  IDEEPTensorId,
+
+  // Here are backends which specify more specialized operators
+  // based on the dtype of the tensor.
+  QuantizedCPUTensorId, // registered at build/aten/src/ATen/QuantizedCPUType.cpp
+  ComplexCPUTensorId,   // lives out of tree at https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex
+  ComplexCUDATensorId,  // and https://gitlab.com/pytorch-complex/pytorch-cuda-strided-complex
+                        // tested at test/cpp_extensions/complex_registration_extension.cpp
+                        // TODO: Remove Complex dispatch keys when Complex is moved in tree
+
+  // This backend is to support custom RNGs; it lets you go
+  // to a different kernel if you pass in a generator that is not a
+  // traditional CPUGenerator/CUDAGenerator.  To make use of this
+  // key:
+  //  1) set it as a second parameter of at::Generator constructor call in
   //     the user-defined PRNG class.
-  //  2) as a dispatch key while registering custom kernels
+  //  2) use it as a dispatch key while registering custom kernels
   //     (templatized kernels specialized for user-defined PRNG class)
+  // intended for out of tree use; tested by aten/src/ATen/test/rng_test.cpp
   CustomRNGKeyId,
 
+  // Here are backends which specify more specialized operators
+  // based on the layout of the tensor.  Note that the sparse backends
+  // are one case where ordering matters: sparse multi-dispatches with
+  // the corresponding dense tensors, and must be handled before them.
+  MkldnnCPUTensorId,  // registered at build/aten/src/ATen/MkldnnCPUType.cpp
+                      // NB: not to be confused with MKLDNNTensorId, which is Caffe2 only
+  SparseCPUTensorId,  // registered at build/aten/src/ATen/SparseCPUType.cpp
+  SparseCUDATensorId, // registered at build/aten/src/ATen/SparseCUDAType.cpp
+  SparseHIPTensorId,  // TODO: I think this is not actually used, due to Note [Masquerading as CUDA]
+
+  // Here are reserved backends for user-defined backends, see Note [Private use TensorId]
+  // To see some example about how to use this, check out MSNPU
+  PrivateUse1_TensorId,
+  PrivateUse2_TensorId,
+  PrivateUse3_TensorId,
+
+
+
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~ AUTOGRAD ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
+  // All backends are oblivious to autograd; autograd is handled as a
+  // layer which happens on top of all backends.  It inspects the autograd
+  // metadata of all inputs, determines what autograd metadata should be
+  // constructed by the output, and otherwise defers to the backend to
+  // actually do the numeric computation.
```

Comment:
Should this comment be after the `VariableTensorId`? Or, at least, there should be some info about `VariableTensorId` it self

Reply:
Added

```
VariableTensorId contains
    // the bulk of this logic.
```